src/ray/object_manager/plasma/create_request_queue.cc

```diff
@@ -88,50 +88,38 @@ Status CreateRequestQueue::ProcessRequest(std::unique_ptr<CreateRequest> &reques
     // Always try to evict after the first attempt.
     evict_if_full = true;
   }
-
   request->error = request->create_callback(evict_if_full, &request->result);
-  Status status;
-  auto should_retry_on_oom = max_retries_ == -1 || num_retries_ < max_retries_;
-  if (request->error == PlasmaError::TransientOutOfMemory) {
-    // The object store is full, but we should wait for space to be made
-    // through spilling, so do nothing. The caller must guarantee that
-    // ProcessRequests is called again so that we can try this request again.
-    // NOTE(swang): There could be other requests behind this one that are
-    // actually serviceable. This may be inefficient, but eventually this
-    // request will get served and unblock the following requests, once
-    // enough objects have been spilled.
-    // TODO(swang): Ask the raylet to spill enough space for multiple requests
-    // at once, instead of just the head of the queue.
-    num_retries_ = 0;
-    status =
-        Status::TransientObjectStoreFull("Object store full, queueing creation request");
-  } else if (request->error == PlasmaError::OutOfMemory && should_retry_on_oom) {
-    num_retries_++;
-    RAY_LOG(DEBUG) << "Not enough memory to create the object, after " << num_retries_
-                   << " tries";
-
-    if (trigger_global_gc_) {
-      trigger_global_gc_();
-    }
-
-    status = Status::ObjectStoreFull("Object store full, should retry on timeout");
-  } else if (request->error == PlasmaError::OutOfMemory) {
-    RAY_LOG(ERROR) << "Not enough memory to create object " << request->object_id
-                   << " after " << num_retries_
-                   << " tries, will return OutOfMemory to the client";
-  }
-
-  return status;
+  return request->error == PlasmaError::OK;
 }
 
 Status CreateRequestQueue::ProcessRequests() {
   while (!queue_.empty()) {
     auto request_it = queue_.begin();
-    auto status = ProcessRequest(*request_it);
-    if (status.IsTransientObjectStoreFull() || status.IsObjectStoreFull()) {
-      return status;
+    auto create_ok = ProcessRequest(*request_it);
+    auto now = absl::GetCurrentTimeNanos();
+    if (create_ok) {
+      FinishRequest(request_it);
+      last_success_ns_ = now;
+    } else {
+      if (trigger_global_gc_) {
+        trigger_global_gc_();
+      }
+      if (spill_objects_callback_()) {
+        last_success_ns_ = absl::GetCurrentTimeNanos();
+        return Status::TransientObjectStoreFull("Waiting for spilling.");
+      } else if (now - last_success_ns_ < 5e9) {
+        // TODO(ekl) make this threshold a ray config def.
+        // We need a grace period since (1) global GC takes a bit of time to
+        // kick in, and (2) there is a race between spilling finishing and space
+        // actually freeing up in the object store.
+        num_retries_ += 1;
```

Comment:
@stephanie-wang Do you think it is okay to just keep this logic instead of retry? Lmk if you think we should change it.